validator-generator/src/main/java/io/avaje/validation/generator/ElementAnnotationContainer.java

```diff
@@ -2,156 +2,131 @@
+            .map(
+                a ->
+                    Map.entry(
+                        UType.parse(a.getAnnotationType()),
+                        AnnotationUtil.annotationAttributeMap(a, element)))
+            .toList();
 
     final var annotations =
-      element.getAnnotationMirrors().stream()
-        .filter(m -> !ValidPrism.isInstance(m))
-        .filter(ElementAnnotationContainer::hasMetaConstraintAnnotation)
-        .map(a -> {
-          if (CrossParamConstraintPrism.isPresent(a.getAnnotationType().asElement())) {
-            crossParam.put(
-              UType.parse(a.getAnnotationType()),
-              AnnotationUtil.annotationAttributeMap(a, element));
-            return null;
-          }
-          return a;
-        })
-        .filter(Objects::nonNull)
-        .collect(
-          toMap(
-            a -> UType.parse(a.getAnnotationType()),
-            a -> AnnotationUtil.annotationAttributeMap(a, element)));
+        Stream.concat(element.getAnnotationMirrors().stream(), uType.annotations().stream())
+            .filter(m -> !ValidPrism.isInstance(m))
+            .filter(ElementAnnotationContainer::hasMetaConstraintAnnotation)
+            .map(
+                a -> {
+                  if (CrossParamConstraintPrism.isPresent(a.getAnnotationType().asElement())) {
+                    crossParam.add(
+                        Map.entry(
+                            UType.parse(a.getAnnotationType()),
+                            AnnotationUtil.annotationAttributeMap(a, element)));
+                    return null;
+                  }
+                  return a;
+                })
+            .filter(Objects::nonNull)
+            .map(
+                a ->
+                    Map.entry(
+                        UType.parse(a.getAnnotationType()),
+                        AnnotationUtil.annotationAttributeMap(a, element)))
+            .distinct()
+            .collect(toList());
 
     if (Util.isNonNullable(element)) {
       var nonNull = UType.parse(APContext.typeElement(NonNullPrism.PRISM_TYPE).asType());
-      annotations.put(nonNull, "Map.of(\"message\",\"{avaje.NotNull.message}\")");
+      annotations.add(Map.entry(nonNull, "Map.of(\"message\",\"{avaje.NotNull.message}\")"));
     }
 
-    return new ElementAnnotationContainer(uType, hasValid, annotations, typeUse1, typeUse2, crossParam);
+    return new ElementAnnotationContainer(
+        uType, hasValid, annotations, typeUse1, typeUse2, crossParam);
   }
 
   static boolean hasMetaConstraintAnnotation(AnnotationMirror m) {
     return hasMetaConstraintAnnotation(m.getAnnotationType().asElement())
-      || ValidPrism.isInstance(m);
+        || ValidPrism.isInstance(m);
   }
 
   static boolean hasMetaConstraintAnnotation(Element element) {
     return ConstraintPrism.isPresent(element);
   }
 
-  // it seems we cannot directly retrieve mirrors from var elements, so var Elements needs special handling
```

Comment:
since we're also using  `uType.annotations()` in the other create constructor this is redundant.